### ISSUE TITLE
Module extension removal issue

### ISSUE THREAD

**Issue by colinbate** [brunch/brunch#1241]
### Description

If you have multiple "extensions" in your filenames (that is multiple periods), an alias is made for that module by removing each in turn. This can cause a different module than the one you were expecting to be imported. This is most likely to occur when dealing with .jsx and .ts(x) files.

For example, if you have `App.jsx` and `App.component.jsx` in the same folder, the latter will sort first and create an alias for `App` which the former won't overwrite because `addExtensions` is only checking for `.js`.

While that example might be a bit contrived, it is becoming popular to use this format of file names to identify types of files. This is also a problem with templates. If you have `some-file.ts` and beside it `some-file.html` which is being compiled to JS, then the alias `some-file` will be taken by the template, not the code module. Probably not what was intended.

### Environment

1. Brunch: 2.5.0
2. Node: 5.7.0
3. NPM: 3.7.5
4. Operating system: Mac OSX 10.11

### `package.json` contents

```json
{
  "name": "your-app",
  "description": "Description",
  "author": "Your Name",
  "version": "0.0.1",
  "repository": {
    "type": "git",
    "url": ""
  },
  "scripts": {
    "start": "brunch watch --server",
    "prod": "brunch build --production"
  },
  "dependencies": {
    "react": "~0.14.7",
    "react-dom": "~0.14.7"
  },
  "devDependencies": {
    "auto-reload-brunch": "^2.0.0",
    "babel-brunch": "~6.0.0",
    "babel-preset-es2015": "~6.3.13",
    "babel-preset-react": "~6.3.13",
    "brunch": "^2.4.0",
    "clean-css-brunch": "^2.0.0",
    "css-brunch": "^2.0.0",
    "javascript-brunch": "^2.0.0",
    "uglify-js-brunch": "^2.0.0"
  }
}
```
### brunch config contents

```js
module.exports = {
  files: {
    javascripts: {
      joinTo: {
        'vendor.js': /^(?!app)/,
        'app.js': /^app/
      }
    },
    stylesheets: {joinTo: 'app.css'}
  },

  plugins: {
    babel: {presets: ['es2015', 'react']}
  }
};
```

### Minimal Reproduction

I started from `brunch\with-react` and added one file: https://github.com/colinbate/with-react/commit/ac473b2f032167ea5ade506b38625b9f520fb062

When you build and run that in the browser, you can see that you no longer import the file you were originally expecting.

**Note:** This only fails on Mac/Linux. On Windows the file list contains both `components/App` and `components\App` and the former sorts before `components\App.component`. It included both files in the bundled files, so I'm going to raise a separate issue for that.

**Comment by goshakkk** [comment]
See this comment that pertains to the issue on the motivation of why it works like this: https://github.com/brunch/commonjs-require-definition/pull/12#issuecomment-186638836

Now, I do agree this probably isn't the best, and I've had to add some hacky checks here and there (and make them `js`-only... :-1:) to make `$.js` that requires `$.a.js` work properly without aliasing `$.a.js` to `$`.

One way I see to resolve that is: by default, drop only the rightmost extension and maybe have a config option that would allow some kinds of extensions to be recursively dropped. Maybe something like:

```javascript
module.exports = {
  modules: {
    recursivelyDrop: ['.tpl.html']
  }
};
```

cc'ing @jacwright since he advocated for the current approach. What do you think of something like this?

**Comment by colinbate** [comment]
I think my concern is that as implemented, this is a breaking change for some apps.

Even for the original use case of being able to include code and a template with the same name is broken without your hack to check for the `.js` extension explicitly. And it is broken for `.jsx` and `.ts` files.

I think technically speaking only the rightmost extension is an extension, I'd consider anything else to be part of the file name. So defaulting to only removing the real extension and providing a way to specify explicit compound extensions to remove would be ideal, and not a breaking change from pre 2.5.0.

**Comment by jacwright** [comment]
@goshakkk, @colinbate, the reason for adding the extension into the definition and aliasing it was to allow `file.js` and `file.html` to be in the same directory. Because templates are added later `file.js` wins. You must require the template using the full filename (including extension).

I think it would reasonable to only drop the last `.abc`, but my concern was that it *would* be a breaking change. Before extensions and aliases were implemented, did `template.tpl.html` become just `template`? Or did it become `template.tpl`? If the latter, then no breaking change here, let's fix this. If the former, then we are indeed breaking existing apps, and I cannot make the call as to whether we should move with that.

Remember, you are never stuck like you were before. You can always require the correct file using a more specific name, such as `App.jsx`.

Here are two better options:
  1. We could add a little more logic to the aliasing. Right now if an alias exists we don't overwrite it. But instead, we could let the shortest number of extensions win. This would mean that `App.jsx` would overwrite `App.component.jsx` and `Object.js` would win out over `Object.assign.js`. This would still let code win over templates. And if you have a folder of `mytemplate.tpl.html` you could still reference by `mytemplate`. The only edge-case here would be something like `file.coffee.js` would lose to `file.html`, which isn't a real case, but I can't think of a real-world instance like this. Can you think of any?
  2. We could let the aliasing be done in brunch (instead of commonjs-require-definition). Brunch knows the extensions for various files, so it can just strip of the "registered extension" and alias it (if the alias doesn't already exist so code still wins over templates).

We should go with the "best" solution, but let it be known you can always use the full filename to require a module, so you are never stuck.

**Comment by jacwright** [comment]
> Even for the original use case of being able to include code and a template with the same name is broken without your hack to check for the .js extension explicitly. And it is broken for .jsx and .ts files.

I don't believe the original use case is broken. Since templates are always added to the file after source code they don't overwrite the alias. But this doesn't solve for two templates (or two code files) with the same base name. I think using option 1 or 2 above would be better. #2 is the best case, but #1 I think is easier and solves 99.9% situations (number taken from an extremely official and reliable source). ;)

**Comment by goshakkk** [comment]
An even easier solution might be to only drop the rightmost extension and let the first one win. If you have more than one level of extensions, just import these explicitly. Keeps brunch simple, also probably less confusing for everyone.

**Comment by goshakkk** [comment]
Fwiw, prior that that PR, we didn't recursively drop extensions. You proposed it and it looked fine on the surface, but it seems now like it really is complicating the matters for everyone :)

>  Before extensions and aliases were implemented, did template.tpl.html become just template? Or did it become template.tpl?

**Comment by colinbate** [comment]
All my templates are first because they sort first alphabetically.

I had `file.tmpl.html` and `file.ts` in my app and the template is what is now imported when `require('file')` is called. Before only one extension was removed and that was your only option as you say.

I'd like to have `file.html` and `file.ts` but even if we only remove one extension with the current sorting the template would still be first. I could require `file.ts` which is ok I suppose.

I suspect your templates are sorting last because you had the `.tpl` in there sorting after `.js`. That won't be the case with just `.html` (or `.jade` or `.ejs` etc)

**Comment by colinbate** [comment]
Also I'm not saying that I'm stuck. All I'm saying is that this was a breaking change. I have to make a bunch of changes to my app to get it to work with 2.5 if I want to update.

**Comment by colinbate** [comment]
@goshakkk I'm liking your idea of just removing a single extension. Keeps the old behaviour. It doesn't solve the `file.html` and `file.js` issue if you want to reference the code without an extension but I think that is fine for now.

**Comment by goshakkk** [comment]
https://github.com/brunch/commonjs-require-definition/pull/20

**Comment by goshakkk** [comment]
Re ordering, I think the simple way to address that would be to change your config to have the following:

```javascript
module.exports = {
  files: {
    templates: {
      order: {
        after: [/\.html$/]
      }
    }
  }
};
```

This will work because javascript and templates are ordered simultaneously

**Comment by jacwright** [comment]
I'm all for just one extension. And @colinbate, I may be wrong, but I thought all templates were added together after all the code, because brunch processes them after it processes all the code. Are you sure your template modules are appearing before your code modules in the final app.js file?

**Comment by jacwright** [comment]
Oh, my page didn't update. Looks like I'm a bit late to the party. Feel free to ignore those comments just now. :smile:

**Comment by goshakkk** [comment]
@paulmillr this issue can be closed.
